phpBB/phpbb/members/controller/online.php

```diff
@@ -0,0 +1,654 @@
+		}
+		$this->db->sql_freeresult($result);
+
+		$this->assign_legend_template_data();
+
+		// Refreshing the page every 60 seconds...
+		meta_refresh(60, $this->helper->route('phpbb_members_online', ['sg' => $show_guests, 'sk' => $sort_key, 'sd' => $sort_dir, 'start' => $start]));
+
+		$start = $this->pagination->validate_start($start, $this->config['topics_per_page'], $counter);
+		$base_url = $this->helper->route('phpbb_members_online', ['sg' => $show_guests, 'sk' => $sort_key, 'sd' => $sort_dir]);
+		$this->pagination->generate_template_pagination($base_url, 'pagination', 'start', $counter, $this->config['topics_per_page'], $start);
+
+		// Send data to template
+		$this->template->assign_vars([
+			'TOTAL_REGISTERED_USERS_ONLINE'	=> $this->language->lang('REG_USERS_ONLINE', (int) $logged_visible_online, $this->language->lang('HIDDEN_USERS_ONLINE', (int) $logged_hidden_online)),
+			'TOTAL_GUEST_USERS_ONLINE'		=> $this->language->lang('GUEST_USERS_ONLINE', (int) $guest_counter),
+
+			'U_SORT_USERNAME'		=> $this->helper->route('phpbb_members_online', ['sg' => (int) $show_guests, 'sk' => 'a', 'sd' => (($sort_key == 'a' && $sort_dir == 'a') ? 'd' : 'a')]),
+			'U_SORT_UPDATED'		=> $this->helper->route('phpbb_members_online', ['sg' => (int) $show_guests, 'sk' => 'b', 'sd' => (($sort_key == 'b' && $sort_dir == 'a') ? 'd' : 'a')]),
+			'U_SORT_LOCATION'		=> $this->helper->route('phpbb_members_online', ['sg' => (int) $show_guests, 'sk' => 'c', 'sd' => (($sort_key == 'c' && $sort_dir == 'a') ? 'd' : 'a')]),
+
+			'U_SWITCH_GUEST_DISPLAY'	=> $this->helper->route('phpbb_members_online', ['sg' => (int) !$show_guests]),
+			'L_SWITCH_GUEST_DISPLAY'	=> ($show_guests) ? $this->language->lang('HIDE_GUESTS') : $this->language->lang('DISPLAY_GUESTS'),
+			'S_SWITCH_GUEST_DISPLAY'	=> ($this->config['load_online_guests']) ? true : false,
+			'S_VIEWONLINE'				=> true,
+		]);
+
+		$this->template->assign_block_vars('navlinks', [
+			'BREADCRUMB_NAME'	=> $this->language->lang('WHO_IS_ONLINE'),
+			'U_BREADCRUMB'		=> $this->helper->route('phpbb_members_online'),
+		]);
+
+		make_jumpbox(append_sid($this->phpbb_root_path . "viewforum." . $this->php_ex));
+
+		// Render
+		return $this->helper->render('viewonline_body.html', $this->language->lang('WHO_IS_ONLINE'));
+	}
+
+	/**
+	 * Given a certain page, it returns the title of the page and a link to it.
+	 * There are 2 strategies to detect the page the user is in:
+	 *  - Try to get the controller route from the router service
+	 *  - And second, if is not possible to get the controller route, the
+	 * page is analyzed by path, analyzing the file that is accessed. This
+	 * is mostly for legacy pages, and should be removed in the future.
+	 *
+	 * @param $session_page
+	 * @param $forum_id
+	 * @return array
+	 * @throws \Exception
+	 */
+	protected function get_location($session_page, $forum_id) // $forum_id can be removed in the future i think https://tracker.phpbb.com/browse/PHPBB3-15434
```

Comment:
This methods can be moved to a viewonline_helper or something in the future to reuse them in other places